File: MPI_OpenMP/LAPLACE/MPI/init_save_mpi.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <mpi.h>

double ind2pos(int i, int n, double L) {
    double ret_val;
    ret_val = ((i-1)-(n-1)/ 2.0)*L/(n-1);
    return ret_val;
}

int init_field(double *temp, int n, int L, int ix_start, int iy_start, int ix_size, int iy_size) {
    // initialize the T field
    int ix,iy;
    double x, y;
    const double sigma = 0.1;
    const double tmax = 100.;
    
    for(iy=0;iy<=iy_size+1;++iy)
        for(ix=0;ix<=ix_size+1;++ix){
            x=ind2pos(ix+ix_start, n, L);
            y=ind2pos(iy+iy_start, n, L);
            temp[ix*(iy_size + 2)+iy] = tmax*exp((-((x*x)+(y*y)))/(2.0*(sigma*sigma)));
        }
    return 0;
}

File: MPI_OpenMP/LAPLACE/MPI/init_save_mpi.f90
function ind2pos(i,n,L)
integer, parameter :: dp=kind(1.d0)
integer :: i, n
real(dp) :: ind2pos, L
ind2pos = ((i-1)-(n-1)/ 2.0)*L/(n-1)
end function ind2pos

subroutine init_field(temp,n,L,ix_start,iy_start,ix_size,iy_size)
! initialize the T field
integer, parameter :: dp=kind(1.d0)
integer :: ix,iy,n,ix_start, iy_start, ix_size,iy_size
real(dp), dimension(0:ix_size+1,0:iy_size+1) :: temp
real(dp) :: ind2pos, L, x, y
real(dp), parameter :: sigma = 0.1d0
real(dp), parameter :: tmax = 100.d0

do iy=0,iy_size+1
    do ix=0,ix_size+1
        x = ind2pos(ix+ix_start,n,L)
        y = ind2pos(iy+iy_start,n,L)
        temp(ix,iy) = tmax*exp(-(x**2+y**2)/(2.0*sigma**2))
    enddo
enddo
end subroutine init_field

